### PR TITLE
Catch IOException when closing HttpUrlConnection in UrlResolutionTask

### PR DESCRIPTION
**Pull request by Amokrane**
If an IOException is thrown, the resolution of the URL fails.

### CODE REVIEW

**Comment by Amokrane** [comment]
Hello,

We (Criteo) noticed that when using App Biding with the MoPub SDK some URLs failed to resolve properly. As a consequence, clicking on certain links didn't work on interstitials.

On the console we had the following error:
`     MoPub  I  [com.mopub.common.UrlHandler][failUrlHandling] SDK Error Log - Task for resolving url was cancelled, null `

After some debugging, we were able to find the root cause: cleaning the HttpUrlConnection in `UrlResolutionTask#getRedirectionLocation` was throwing an `IOException` that was not being caught.

I tested locally and this fixes the issue we were seeing.

**Comment by chauduyphanvu** [comment]
@Amokrane, thanks for letting us know! 

Could you share a clickthrough URL (or better yet, the ad unit ID) so we can reproduce and debug this issue? Could you also confirm that handling the IOException allows the clickthrough to happen (rather than consume the error and fail gracefully)?

Vu Chau
MoPub Team

**Comment by Amokrane** [comment]
Here is one URL that used to consistently fail: `https://www.laredoute.fr/pplp/100/97172/161890/cat-161892.aspx?cod=CMG00126703FR&ectrans=1`

AdUnitId -> `83a2996696284da881edaf1a480e5d7c`

Yes, I confirm that the clickthrough does happen properly with this fix.


83a2996696284da881edaf1a480e5d7c
83a2996696284da881edaf1a480e5d7c
83a2996696284da881edaf1a480e5d7c
83a2996696284da881edaf1a480e5d7c
83a2996696284da881edaf1a480e5d7c

**Comment by chauduyphanvu** [comment]
@Amokrane, thanks for the confirmation.

I wasn't able to get fills from `83a2996696284da881edaf1a480e5d7c`. Were any recent changes to it? If not, are you able to share any other ad unit ID? 

```
2020-05-29 18:56:37.878 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.mobileads.AdViewController][loadNonJavascript] SDK Log - Loading url: https://ads.mopub.com/m/ad?v=6&id=83a2996696284da881edaf1a480e5d7c&nv=5.12.0&dn=Google%2CPixel%203%20XL%2Ccrosshatch&bundle=com.mopub.simpleadsdemo&ll=40.018%2C-75.1219&lla=2000&llf=434953&llsdk=1&z=-0400&o=p&cw=1440&ch=2789&w=1440&h=2960&sc=3.5&mcc=311&mnc=480&cn=Verizon&ct=2&av=external&abt=%7B%22tapjoy%22%3A%7B%22token%22%3A%221%22%7D%2C%22facebook%22%3A%7B%22token%22%3A%22eJx1VG1vmzwU%5C%2FSuIT5uUEBsMhPWTSUiDSoABaTeNCRkwHSoJEZCu69T%5C%2F%5C%2FlxD%2BqJHWqRIvq8%2BPvdw%5C%2F8o0DO29663lL7JpyTP5lnohjeguBsffVK77jB%5C%2FOqfwllSvW9DyVZ6nMTg9ZXz%5C%2Fz0Y1VC%2BlEVcfIUB%5C%2F4c3vkWVtVPR%5C%2FGhDleovF3qT1l%5C%2FcC6gZdZx1nfHsckj%2B791daJsk2w99cZDV0Vj%5C%2FkAoOT5%2Bf6e5c104dCdJxijP3tkzfl%5C%2FgfxcN2U2%5C%2FDlN%5C%2FjEvlV%5C%2FgdXvfTb7Dy8a3gE1XKyeOXdv1wJ85PrU9Z%5C%2F0xHoY%2B3Tng2bXhOReeTQiWZmDLrCqiFqTKc4tYBVoaRlGalmFaCGNIjEXjwM8Sd6zHuoVMS7WWhmIRBPEdvRH%2B67a9b8RVKxpFrhOB65Z3NZAoeqxvshUN6YQPQj%2B0GZnpM3OG8QwbM2z%2BhKy148dTWFN0UbWKHMfPto57vU3EXMny0uvWiQQm8OmKrggUYRQIBqZhn7q24H2fHdlhYq5oD8qhPZ1zpa8Pp4azsi%5C%2F5oX0dDJjZpQbSR84gUDQ1Pw5Zfz6d2m54i7y8I7tz18kWriQjUe46i4N9tBJsTDjhwI5l19Yl2EEsUAsG0WjBib4FqU%2B94HpCX%5C%2Fxi3X19vJ9UOYntkdWNEE52ANjdnylEiGlYRNOMSbLtwJqPcc3UCdF1RCbB5mwY%2BGspQsrkrTrOs%5C%2F7EiomopbVEBsbIVNFFaDFomV47fjJq57luGpYugHPpk1cfz09X0uUREkZXUlg%5C%2F8UbSpG%2BeZAvtpouvX1WqqAjpCCsI%5C%2FrZ6Jf1%2B%5C%2FCzRE4zhjuc39QDtNFPRDOnTzTbZeTOpqR%2B4dM2Lh%5C%2FazBBrqa%5C%2Fi2FgSuXP3qWpjoYqkpSCEYacpSk3ZtXjdcilnFuvqt14%2BNTf10Qc8lDLHgPh9%2Bt93Dpu0ucK82dizi79ZtuoAnQJmdLv6hFxGFNP4ERB5ZA6YNpmnB4TYWrAAvcPZWkHPM9rEQdRQEyfgpConsgrXjwfmdJ3DasCg8IZp%5C%2F3CrkEW%2BpEAyzLM3QiY4NVmiYwXArzUBFkZsq03IDI0wsvKzwFdR4wYqObUco4PCd5C6IbrLke%2Bhc8LjrDRUX55iruWHOEbL0OTEIn1uWjuawD3I9N1TE9UqgSJLItfeJ2AbuelI5LJaLrF9ZEW39OKGeNy6BDxtkrMEs5xUi%2Bhxxs5wTgD7PDYPPS4KqpUksC%5C%2Faw%5C%2FPIfBw6uZw%3D%3D%22%7D%2C%22mintegral%22%3A%7B%22token%22%3A%226deIYg5sYde%5C%2FJkes6deIidefNVe6ini%5C%2Fio9ei%2BMM4aSI6deIinVPiahPiU3PfAlbfdMeWUQIidMM6dM2idMM6ajeGniFfbM9GnlMfN%3D%3D%22%7D%2C%22adcolony%22%3A%7B%22token%22%3A%221%22%7D%7D&udid=mp_tmpl_advertising_id&dnt=mp_tmpl_do_not_track&mid=mp_tmpl_mopub_id&gdpr_applies=0&force_gdpr_applies=0&current_consent_status=unknown&backoff_ms=5000&backoff_reason=unfilled&mr=1&vv=3
2020-05-29 18:56:37.883 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.network.AdLoader][fetchAd] Ad requesting from AdServer: https://ads.mopub.com/m/ad
    {"ll":"40.018,-75.1219","vv":"3","nv":"5.12.0","mid":"2144fe4b-d230-4ca3-a4fb-787e2f749764","dn":"Google,Pixel 3 XL,crosshatch","mcc":"311","backoff_ms":"5000","sc":"3.5","current_consent_status":"unknown","id":"83a2996696284da881edaf1a480e5d7c","backoff_reason":"unfilled","udid":"ifa:cb1e2b67-0095-464e-9950-949b5b620e5f","bundle":"com.mopub.simpleadsdemo","llsdk":"1","gdpr_applies":"0","lla":"2000","mnc":"480","ch":"2789","mr":"1","llf":"434953","h":"2960","force_gdpr_applies":"0","cn":"Verizon","dnt":"0","o":"p","ct":"2","abt":"{\"tapjoy\":{\"token\":\"1\"},\"facebook\":{\"token\":\"eJx1VG1vmzwU\\\/SuIT5uUEBsMhPWTSUiDSoABaTeNCRkwHSoJEZCu69T\\\/\\\/lxD+qJHWqRIvq8+Pvdw\\\/8o0DO29663lL7JpyTP5lnohjeguBsffVK77jB\\\/OqfwllSvW9DyVZ6nMTg9ZXz\\\/z0Y1VC+lEVcfIUB\\\/4c3vkWVtVPR\\\/GhDleovF3qT1l\\\/cC6gZdZx1nfHsckj+791daJsk2w99cZDV0Vj\\\/kAoOT5+f6e5c104dCdJxijP3tkzfl\\\/gfxcN2U2\\\/DlN\\\/jEvlV\\\/gdXvfTb7Dy8a3gE1XKyeOXdv1wJ85PrU9Z\\\/0xHoY+3Tng2bXhOReeTQiWZmDLrCqiFqTKc4tYBVoaRlGalmFaCGNIjEXjwM8Sd6zHuoVMS7WWhmIRBPEdvRH+67a9b8RVKxpFrhOB65Z3NZAoeqxvshUN6YQPQj+0GZnpM3OG8QwbM2z+hKy148dTWFN0UbWKHMfPto57vU3EXMny0uvWiQQm8OmKrggUYRQIBqZhn7q24H2fHdlhYq5oD8qhPZ1zpa8Pp4azsi\\\/5oX0dDJjZpQbSR84gUDQ1Pw5Zfz6d2m54i7y8I7tz18kWriQjUe46i4N9tBJsTDjhwI5l19Yl2EEsUAsG0WjBib4FqU+94HpCX\\\/xi3X19vJ9UOYntkdWNEE52ANjdnylEiGlYRNOMSbLtwJqPcc3UCdF1RCbB5mwY+GspQsrkrTrOs\\\/7EiomopbVEBsbIVNFFaDFomV47fjJq57luGpYugHPpk1cfz09X0uUREkZXUlg\\\/8UbSpG+eZAvtpouvX1WqqAjpCCsI\\\/rZ6Jf1+\\\/CzRE4zhjuc39QDtNFPRDOnTzTbZeTOpqR+4dM2Lh\\\/azBBrqa\\\/i2FgSuXP3qWpjoYqkpSCEYacpSk3ZtXjdcilnFuvqt14+NTf10Qc8lDLHgPh9+t93Dpu0ucK82dizi79ZtuoAnQJmdLv6hFxGFNP4ERB5ZA6YNpmnB4TYWrAAvcPZWkHPM9rEQdRQEyfgpConsgrXjwfmdJ3DasCg8IZp\\\/3CrkEW+pEAyzLM3QiY4NVmiYwXArzUBFkZsq03IDI0wsvKzwFdR4wYqObUco4PCd5C6IbrLke+hc8LjrDRUX55iruWHOEbL0OTEIn1uWjuawD3I9N1TE9UqgSJLItfeJ2AbuelI5LJaLrF9ZEW39OKGeNy6BDxtkrMEs5xUi+hxxs5wTgD7PDYPPS4KqpUksC\\\/aw\\\/PIfBw6uZw==\"},\"mintegral\":{\"token\":\"6deIYg5sYde\\\/Jkes6deIidefNVe6ini\\\/io9ei+MM4aSI6deIinVPiahPiU3PfAlbfdMeWUQIidMM6dM2idMM6ajeGniFfbM9GnlMfN==\"},\"adcolony\":{\"token\":\"1\"}}","cw":"1440","av":"external","v":"6","w":"1440","z":"-0400"}
2020-05-29 18:56:37.960 20984-21130/com.mopub.simpleadsdemo I/MoPub: [com.mopub.network.MultiAdResponse][parseSingleAdResponse] Ad server responded with:
    {"content":" <!DOCTYPE html> <html> <head>  <!-- Adgroup is 244cc71615e144528c75aaed572d4a09 -->  <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0, user-scalable=no\">  <style type='text\/css'> .mp_center { position: fixed; top: 50%; left: 50%; margin-left: -160px !important; margin-top: -240px !important; } <\/style>  <script type=\"text\/javascript\"> function mopubFinishLoad(){ setTimeout(function() { window.location = 'mopub:\/\/finishLoad'; }, 0); } <\/script>  <script type=\"text\/javascript\"> function webviewDidClose(){ if (typeof webviewDidCloseHelper == 'function') { webviewDidCloseHelper(); } } function webviewDidAppear(){  if(typeof trackImpressionHelper == 'function') { trackImpressionHelper(); }  if(typeof webviewDidAppearHelper == 'function') { webviewDidAppearHelper(); } } window.addEventListener(\"load\", function() { var links = document.getElementsByTagName('a'); for(var i=0; i < links.length; i++) { links[i].setAttribute('target','_blank'); } }, false);  <\/script>  <\/head> <body style=\"margin:0;padding:0;\"> <script type=\"text\/javascript\">\n    function write(html) {\n        document.open();\n        document.write(html);\n        document.close();\n    }\n\n    function mopubFailover() {\n        loaded=true; window.location=\"mopub:\/\/failLoad\";\n    }\n\n    var url = \"\";\n\n    if (url.length > 0) {\n        write(\"<scr\" + \"ipt type='text\/javascript' src='\" + url + \"'><\/scr\" + \"ipt>\");\n    } else {\n        mopubFailover();\n    }\n<\/script> <script type=\"text\/javascript\"> if (typeof htmlWillCallFinishLoad == \"undefined\" || !htmlWillCallFinishLoad) { if(typeof mopubFinishLoad == 'function') { window.onload = mopubFinishLoad; } }  <\/script>  <\/body> <\/html> ","metadata":{"allow-custom-close":1,"content-type":"text\/html; charset=UTF-8","imptrackers":["https:\/\/ads.mopub.com\/m\/imp?appid=&cid=64b7b63b7bfe47edbc2281dc8fb862b3&city=Philadelphia&ckv=2&country_code=US&cppck=13105&current_consent_status=unknown&dev=Pixel%203%20XL&exclude_adgroups=244cc71615e144528c75aaed572d4a09&gdpr_applies=0&id=83a2996696284da881edaf1a480e5d7c&is_mraid=0&os=Android&osv=10.0.0&priority=5&req=7fb3558dcea940899b4e44c2f0971bc9_00e77ef40005b22d&reqt=1590792998.0&rev=0.020000&udid=ifa%3Acb1e2b67-0095-464e-9950-949b5b620e5f&video_type="],"vast-player-version":1,"x-ad-timeout-ms":30000,"x-adgroupid":"244cc71615e144528c75aaed572d4a09","x-adtype":"html","x-before-load-url":"https:\/\/ads.mopub.com\/m\/attempt?account_id=350acfac9dbf4beb9dfe300dad210e19&adgroup_id=244cc71615e144528c75aaed572d4a09&adgroup_priority=5&adgroup_type=gtee&campaign_id=9afb7140f6494055a65b543865a99072&cid=64b7b63b7bfe47edbc2281dc8fb862b3&country_code=US&cppck=91693&current_consent_status=unknown&gdpr_applies=0&id=83a2996696284da881edaf1a480e5d7c&logged_server_side=0&network_external_key=&network_id=&req=7fb3558dcea940899b4e44c2f0971bc9_00e77ef40005b22d&reqt=1590792998.0&rtc=0&segment_id=","x-browser-agent":0,"x-clickthrough":"https:\/\/ads.mopub.com\/m\/aclk?appid=&cid=
2020-05-29 18:56:37.960 20984-21130/com.mopub.simpleadsdemo I/MoPub: [com.mopub.network.MultiAdResponse][parseSingleAdResponse] 64b7b63b7bfe47edbc2281dc8fb862b3&city=Philadelphia&ckv=2&country_code=US&cppck=606DC&current_consent_status=unknown&dev=Pixel%203%20XL&exclude_adgroups=244cc71615e144528c75aaed572d4a09&gdpr_applies=0&id=83a2996696284da881edaf1a480e5d7c&is_mraid=0&os=Android&osv=10.0.0&priority=5&req=7fb3558dcea940899b4e44c2f0971bc9_00e77ef40005b22d&reqt=1590792998.0&rev=0&udid=ifa%3Acb1e2b67-0095-464e-9950-949b5b620e5f&video_type=","x-creativeid":"64b7b63b7bfe47edbc2281dc8fb862b3","x-orientation":"p","x-scrollable":"0"}}
2020-05-29 18:56:37.964 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.mobileads.MoPubInterstitial$MoPubInterstitialView][loadCustomEvent] Ad Log - Loading custom event interstitial adapter.
2020-05-29 18:56:37.966 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.mobileads.CustomEventInterstitialAdapter][<init>] SDK Log - Attempting to invoke custom event: com.mopub.mobileads.HtmlInterstitial
2020-05-29 18:56:37.967 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.mobileads.CustomEventInterstitialAdapter][loadInterstitial] SDK Log - loadInterstitial()
2020-05-29 18:56:37.967 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.mobileads.ResponseBodyInterstitial][loadInterstitial] Ad attempting to load
2020-05-29 18:56:37.968 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.mobileads.HtmlInterstitial][preRenderHtml] Adapter HtmlInterstitial attempting to load ad
2020-05-29 18:56:37.969 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.mobileads.MoPubActivity][preRenderHtml] Ad attempting to load
2020-05-29 18:56:37.989 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.common.ExternalViewabilitySessionManager][logEvent] SDK Log - AVID viewability event: start display session.
2020-05-29 18:56:37.991 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.common.ExternalViewabilitySessionManager][logEvent] SDK Log - Moat viewability event: start display session.
2020-05-29 18:56:37.998 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.mobileads.ResponseBodyInterstitial][loadInterstitial] Ad loaded
2020-05-29 18:56:38.010 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.mobileads.CustomEventInterstitialAdapter][onInterstitialFailed] SDK Log - onInterstitialFailed() failed with code 10000 and message Unspecified error.
2020-05-29 18:56:38.011 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.mobileads.MoPubInterstitial][onCustomEventInterstitialFailed] Ad failed to load: (10,000) Unspecified error.
2020-05-29 18:56:38.012 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.mobileads.AdViewController][loadFailUrl] Ad failed to load: (10,000) Unspecified error.
2020-05-29 18:56:38.012 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.mobileads.AdViewController][loadNonJavascript] SDK Log - Loading url: 
2020-05-29 18:56:38.017 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.network.AdLoader][fetchAd] Ad requesting from AdServer: https://ads.mopub.com/m/ad
    {"ll":"40.018,-75.1219","vv":"3","nv":"5.12.0","mid":"2144fe4b-d230-4ca3-a4fb-787e2f749764","dn":"Google,Pixel 3 XL,crosshatch","backoff_ms":"5000","mcc":"311","sc":"3.5","current_consent_status":"unknown","tqr":"XtGTJj65mLk6Ov2XD6PFpsYLqssyklsNySwvTA","exclude":"244cc71615e144528c75aaed572d4a09","backoff_reason":"unfilled","id":"83a2996696284da881edaf1a480e5d7c","udid":"ifa:cb1e2b67-0095-464e-9950-949b5b620e5f","bundle":"com.mopub.simpleadsdemo","gdpr_applies":"0","llsdk":"1","lla":"2000","rtc":"1","ac":"0","mnc":"480","ch":"2789","mr":"1","llf":"434953","force_gdpr_applies":"0","h":"2960","cn":"Verizon","dnt":"0","cppck":"62DC2","o":"p","ct":"2","abt":"{\"tapjoy\":{\"token\":\"1\"},\"facebook\":{\"token\":\"eJx1VG1vmzwU\\\/SuIT5uUEBsMhPWTSUiDSoABaTeNCRkwHSoJEZCu69T\\\/\\\/lxD+qJHWqRIvq8+Pvdw\\\/8o0DO29663lL7JpyTP5lnohjeguBsffVK77jB\\\/OqfwllSvW9DyVZ6nMTg9ZXz\\\/z0Y1VC+lEVcfIUB\\\/4c3vkWVtVPR\\\/GhDleovF3qT1l\\\/cC6gZdZx1nfHsckj+791daJsk2w99cZDV0Vj\\\/kAoOT5+f6e5c104dCdJxijP3tkzfl\\\/gfxcN2U2\\\/DlN\\\/jEvlV\\\/gdXvfTb7Dy8a3gE1XKyeOXdv1wJ85PrU9Z\\\/0xHoY+3Tng2bXhOReeTQiWZmDLrCqiFqTKc4tYBVoaRlGalmFaCGNIjEXjwM8Sd6zHuoVMS7WWhmIRBPEdvRH+67a9b8RVKxpFrhOB65Z3NZAoeqxvshUN6YQPQj+0GZnpM3OG8QwbM2z+hKy148dTWFN0UbWKHMfPto57vU3EXMny0uvWiQQm8OmKrggUYRQIBqZhn7q24H2fHdlhYq5oD8qhPZ1zpa8Pp4azsi\\\/5oX0dDJjZpQbSR84gUDQ1Pw5Zfz6d2m54i7y8I7tz18kWriQjUe46i4N9tBJsTDjhwI5l19Yl2EEsUAsG0WjBib4FqU+94HpCX\\\/xi3X19vJ9UOYntkdWNEE52ANjdnylEiGlYRNOMSbLtwJqPcc3UCdF1RCbB5mwY+GspQsrkrTrOs\\\/7EiomopbVEBsbIVNFFaDFomV47fjJq57luGpYugHPpk1cfz09X0uUREkZXUlg\\\/8UbSpG+eZAvtpouvX1WqqAjpCCsI\\\/rZ6Jf1+\\\/CzRE4zhjuc39QDtNFPRDOnTzTbZeTOpqR+4dM2Lh\\\/azBBrqa\\\/i2FgSuXP3qWpjoYqkpSCEYacpSk3ZtXjdcilnFuvqt14+NTf10Qc8lDLHgPh9+t93Dpu0ucK82dizi79ZtuoAnQJmdLv6hFxGFNP4ERB5ZA6YNpmnB4TYWrAAvcPZWkHPM9rEQdRQEyfgpConsgrXjwfmdJ3DasCg8IZp\\\/3CrkEW+pEAyzLM3QiY4NVmiYwXArzUBFkZsq03IDI0wsvKzwFdR4wYqObUco4PCd5C6IbrLke+hc8LjrDRUX55iruWHOEbL0OTEIn1uWjuawD3I9N1TE9UqgSJLItfeJ2AbuelI5LJaLrF9ZEW39OKGeNy6BDxtkrMEs5xUi+hxxs5wTgD7PDYPPS4KqpUksC\\\/aw\\\/PIfBw6uZw==\"},\"mintegral\":{\"token\":\"6deIYg5sYde\\\/Jkes6deIidefNVe6ini\\\/io9ei+MM4aSI6deIinVPiahPiU3PfAlbfdMeWUQIidMM6dM2idMM6ajeGniFfbM9GnlMfN==\"},\"adcolony\":{\"token\":\"1\"}}","cw":"1440","av":"external","v":"6","w":"1440","z":"-0400","request_id":"7fb3558dcea940899b4e44c2f0971bc9_00e77ef40005b22d"}
2020-05-29 18:56:38.017 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.network.TrackingRequest$1][onResponse] SDK Log - Successfully hit tracking endpoint: https://ads.mopub.com/m/attempt
2020-05-29 18:56:38.095 20984-21131/com.mopub.simpleadsdemo I/MoPub: [com.mopub.network.MultiAdResponse][parseSingleAdResponse] Ad server responded with:
    {"content":"","metadata":{"x-adtype":"clear","x-backfill":"clear","x-refreshtime":60}}
2020-05-29 18:56:38.096 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.network.AdLoader$1][onErrorResponse] Ad server responded with:
    No ads found for ad unit.
2020-05-29 18:56:38.096 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.mobileads.AdViewController][adDidFail] SDK Log - Ad failed to load.
2020-05-29 18:57:55.000 20984-20984/com.mopub.simpleadsdemo I/MoPub: [com.mopub.mobileads.MoPubView][destroy] Ad Log - Destroy() called
```

**Comment by Amokrane** [comment]
@chauduyphanvu thanks for looking into this. I wasn't clear at the beginning, but we are testing this in the context of our SDK which get fills from our own backend and then inject the displayUrl on the MoPub interstitial/banners.

I dug into this more and I think I know which https URLs end up failing. In the case of LaRedoute, they use DataDome (bot protection) and a basic call to getResponseCode() (https://github.com/mopub/mopub-android-sdk/blob/master/mopub-sdk/mopub-sdk-base/src/main/java/com/mopub/common/UrlResolutionTask.java#L121) returns a 403 error.

This can be easily tested with this curl request:
`curl --location --head 'https://www.laredoute.fr/pplp/100/97172/161890/cat-161892.aspx?cod=CMG00126703FR&ectrans=1' 
`
which returns:

```
HTTP/1.1 403 Forbidden
Cache-Control: private
Pragma: no-cache
Content-Type: text/html; charset=utf-8
X-DataDome: protected
...
```
I think that all advertisers that end up having such kind of protection could experience a zero landing rate. Opening the URL in a browser (like MoPub's in-app browser, a WebView or any browser) works and doesn't return a 403, probably because the right set of headers are automatically added. If I am not mistaken, HttpUrlConnection doesn't add any headers by default.

The PR that I am suggesting here does not make that request return a 2xx or 3xx response, but by properly closing the HTTP connection in the finally clause (https://github.com/mopub/mopub-android-sdk/blob/master/mopub-sdk/mopub-sdk-base/src/main/java/com/mopub/common/UrlResolutionTask.java#L103 -> this could throw IOException), it ensures the graceful termination of URLResolutionTask and return the last non-null URL (which ends up being the one that returned 403) to the URLHandler (https://github.com/mopub/mopub-android-sdk/blob/master/mopub-sdk/mopub-sdk-base/src/main/java/com/mopub/common/UrlResolutionTask.java#L150). As a consequence the clickthrough happens.

Now, it's unclear from the codebase whether you want URLs that return 4xx during the resolution process to fail, and therefore the clickthrough not to happen. For instance, why don't you check 4xx response codes here: https://github.com/mopub/mopub-android-sdk/blob/master/mopub-sdk/mopub-sdk-base/src/main/java/com/mopub/common/UrlResolutionTask.java#L125?

To ease the reproduction of this on your side, I have extracted the relevant code in a separate repo: https://github.com/Amokrane/MopubLaRedouteLink. The relevant line is here: https://github.com/Amokrane/MopubLaRedouteLink/blob/master/app/src/main/java/com/chentir/mopublaredoutelink/MainActivity.java#L70 (this will be caught and a log will be printed on the console).

I don't know if you have any kind of metrics on your end to measure landing rates, but it will be interesting to measure the impact of this fix (if it ever gets merged) or any other fix (such as setting the right headers during URL resolution), on that metric.

In any case, I will be happy to provide more details and exchange more if needed.

Cheers.
